### PR TITLE
mark all BinaryFormatter-related tests as conditional

### LINKED ISSUE #103406: [android][apple] BinaryFormatter serialization and deserialization not supported on mobile targets
## Build Information
Build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=703011
Build error leg or test failing: System.Resources.Extensions.Tests.FormattedObject
Affected configurations:
- `android_arm/arm64/x64/x86`
- `iossimulator_x64`
- `maccatalyst_arm64/x64`
- `tvos_arm64`
- `tvossimulator_x64`

## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssueJsonStepByStep.md).

```json
{
  "ErrorMessage": "System.PlatformNotSupportedException : BinaryFormatter serialization and deserialization are not supported on this platform.",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

## Stacktrace example
```
[17:48:42.8864320] 2024-06-10 14:48:42.918 System.Resources.Extensions.BinaryFormat.Tests[6435:96157250] 	[FAIL] System.Resources.Extensions.Tests.FormattedObject.MultidimensionalArrayTests.IntegerArrays(lengths: [4, 3, 2])
[17:48:42.8871530] 2024-06-10 14:48:42.919 System.Resources.Extensions.BinaryFormat.Tests[6435:96157250] System.PlatformNotSupportedException : BinaryFormatter serialization and deserialization are not supported on this platform. See https://aka.ms/binaryformatter for more information.
[17:48:42.8871680]    at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream , Object )
[17:48:42.8878060] 2024-06-10 14:48:42.919 System.Resources.Extensions.BinaryFormat.Tests[6435:96157250]    at System.Resources.Extensions.Tests.Common.ISerializer.Serialize(Object value, SerializationBinder binder, ISurrogateSelector surrogateSelector, FormatterTypeStyle typeStyle)
[17:48:42.8884710] 2024-06-10 14:48:42.920 System.Resources.Extensions.BinaryFormat.Tests[6435:96157250]    at System.Resources.Extensions.Tests.Common.SerializationTest`1[[System.Resources.Extensions.Tests.FormattedObject.FormattedObjectSerializer, System.Resources.Extensions.BinaryFormat.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]].Serialize(Object value, SerializationBinder binder, ISurrogateSelector surrogateSelector, FormatterTypeStyle typeStyle)
[17:48:42.8890730] 2024-06-10 14:48:42.921 System.Resources.Extensions.BinaryFormat.Tests[6435:96157250]    at System.Resources.Extensions.Tests.Common.MultidimensionalArrayTests`1[[System.Resources.Extensions.Tests.FormattedObject.FormattedObjectSerializer, System.Resources.Extensions.BinaryFormat.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]].IntegerArrays(Int32[] lengths)
[17:48:42.8898180] 2024-06-10 14:48:42.921 System.Resources.Extensions.BinaryFormat.Tests[6435:96157250]    at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
[17:48:42.8898340]    at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object , Span`1 , BindingFlags )
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=703011
**Error message validated:** `[System.PlatformNotSupportedException : BinaryFormatter serialization and deserialization are not supported on this platform.`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 6/13/2024 9:43:14 AM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[703011](https://dev.azure.com/dnceng-public/public/_build/results?buildId=703011)|dotnet/runtime|[System.Resources.Extensions.BinaryFormat.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=703011&view=ms.vss-test-web.build-test-results-tab&runId=17511622&resultId=203686)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|1|1|

### PR DESCRIPTION
**Pull request by adamsitnik**
fixes #103406

### CODE REVIEW

**Comment by adamsitnik** [comment]
The following failure is related:

```log
  Starting:    System.Resources.Extensions.BinaryFormat.Tests (parallel test collections = on [2 threads], stop on fail = off)
    System.Resources.Extensions.Tests.FormattedObject.SystemDrawingTests.Bitmap_RoundTrip(typeStyle: TypesWhenNeeded, assemblyMatching: Full) [FAIL]
      System.TypeInitializationException : The type initializer for 'System.Drawing.DrawingCom' threw an exception.
      ---- System.PlatformNotSupportedException : Operation is not supported on this platform.
      Stack Trace:
           at System.Drawing.Image.Save(Stream stream, ImageCodecInfo encoder, EncoderParameters encoderParams)
           at System.Drawing.Image.Save(MemoryStream stream)
           at System.Drawing.Image.System.Runtime.Serialization.ISerializable.GetObjectData(SerializationInfo si, StreamingContext context)
```

Once I've added `ConditionalClass` to the base class, the derived type annotation is no longer respected...

```cs
[ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.IsBinaryFormatterSupported))]
public abstract class SerializationTest<TSerializer> where TSerializer : ISerializer

[ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.IsDrawingSupported))]
public abstract class SystemDrawingTests<T> : SerializationTest<T> where T : ISerializer
```

**Comment by adamsitnik** [comment]
Despite adding `[ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsDrawingSupported))]` to `Bitmap_RoundTrip` I keep getting following error for Mono on Windows:

```log
C:\h\w\AA92092E\w\B55809CF\e>"C:\h\w\AA92092E\p\dotnet.exe" exec --runtimeconfig System.Resources.Extensions.BinaryFormat.Tests.runtimeconfig.json --depsfile System.Resources.Extensions.BinaryFormat.Tests.deps.json xunit.console.dll System.Resources.Extensions.BinaryFormat.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
  Discovering: System.Resources.Extensions.BinaryFormat.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Resources.Extensions.BinaryFormat.Tests (found 132 test cases)
  Starting:    System.Resources.Extensions.BinaryFormat.Tests (parallel test collections = on [2 threads], stop on fail = off)
    System.Resources.Extensions.Tests.FormattedObject.SystemDrawingTests.Bitmap_RoundTrip(typeStyle: TypesWhenNeeded, assemblyMatching: Full) [FAIL]
      System.TypeInitializationException : The type initializer for 'System.Drawing.DrawingCom' threw an exception.
      ---- System.PlatformNotSupportedException : Operation is not supported on this platform.
      Stack Trace:
           at System.Drawing.Image.Save(Stream stream, ImageCodecInfo encoder, EncoderParameters encoderParams)
           at System.Drawing.Image.Save(MemoryStream stream)
           at System.Drawing.Image.System.Runtime.Serialization.ISerializable.GetObjectData(SerializationInfo si, StreamingContext context)
        /_/src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectInfo.cs(113,0): at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitSerialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
        /_/src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectInfo.cs(74,0): at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.Serialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
        /_/src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectWriter.cs(87,0): at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Serialize(Object graph, BinaryFormatterWriter serWriter)
        /_/src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatter.Core.cs(83,0): at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream serializationStream, Object graph)
        /_/src/libraries/System.Resources.Extensions/tests/BinaryFormatTests/Common/ISerializer.cs(26,0): at System.Resources.Extensions.Tests.Common.ISerializer.Serialize(Object value, SerializationBinder binder, ISurrogateSelector surrogateSelector, FormatterTypeStyle typeStyle)
        /_/src/libraries/System.Resources.Extensions/tests/BinaryFormatTests/Common/SerializationTest.cs(34,0): at System.Resources.Extensions.Tests.Common.SerializationTest`1[[System.Resources.Extensions.Tests.FormattedObject.FormattedObjectSerializer, System.Resources.Extensions.BinaryFormat.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]].Serialize(Object value, SerializationBinder binder, ISurrogateSelector surrogateSelector, FormatterTypeStyle typeStyle)
        /_/src/libraries/System.Resources.Extensions/tests/BinaryFormatTests/Common/SerializationTest.cs(51,0): at System.Resources.Extensions.Tests.Common.SerializationTest`1[[System.Resources.Extensions.Tests.FormattedObject.FormattedObjectSerializer, System.Resources.Extensions.BinaryFormat.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]].RoundTrip[Bitmap](Bitmap value, SerializationBinder binder, ISurrogateSelector surrogateSelector, FormatterTypeStyle typeStyle, FormatterAssemblyStyle assemblyMatching)
        /_/src/libraries/System.Resources.Extensions/tests/BinaryFormatTests/Common/SystemDrawingTests.cs(20,0): at System.Resources.Extensions.Tests.Common.SystemDrawingTests`1[[System.Resources.Extensions.Tests.FormattedObject.FormattedObjectSerializer, System.Resources.Extensions.BinaryFormat.Tests, Version=9.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51]].Bitmap_RoundTrip(FormatterTypeStyle typeStyle, FormatterAssemblyStyle assemblyMatching)
        /_/src/mono/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.Mono.cs(22,0): at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
        /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(178,0): at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
        ----- Inner Stack Trace -----
```

[full log](https://[helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-103471-merge-427ab9a0850a48cba0/System.Resources.Extensions.BinaryFormat.Tests/1/console.9e66e985.log?helixlogtype=result](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-103471-merge-427ab9a0850a48cba0/System.Resources.Extensions.BinaryFormat.Tests/1/console.9e66e985.log?helixlogtype=result))

@akoeplinger @dotnet/area-system-drawing what am I missing?

**Comment by akoeplinger** [comment]
IsDrawingSupported just checks that we're on Windows, but Mono doesn't support COM Interop which this is apparently trying to use. You can check for SupportsComInterop or IsBuiltInComEnabled

File: src/libraries/System.Resources.Extensions/tests/BinaryFormatTests/Common/SystemDrawingTests.cs [hunk 1]
```diff
@@ -7,10 +7,11 @@
 
 namespace System.Resources.Extensions.Tests.Common;
 
-[ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.IsDrawingSupported))]
+// This type can not be annotated with [ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.IsDrawingSupported))]
+// because the base type is annotated with [ConditionalClass(typeof(PlatformDetection), nameof(PlatformDetection.IsBinaryFormatterSupported))]
 public abstract class SystemDrawingTests<T> : SerializationTest<T> where T : ISerializer
 {
-    [Theory]
+    [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsDrawingSupported))]
```
Comment (adamsitnik) [inline src/libraries/System.Resources.Extensions/tests/BinaryFormatTests/Common/SystemDrawingTests.cs:14]:
```suggestion
    [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsDrawingSupported), nameof(PlatformDetection.SupportsComInterop))]
```

File: src/libraries/System.Resources.Extensions/tests/BinaryFormatTests/Common/SystemDrawingTests.cs [hunk 2]
```diff
@@ -20,7 +21,7 @@ public void Bitmap_RoundTrip(FormatterTypeStyle typeStyle, FormatterAssemblyStyl
         deserialized.Size.Should().Be(bitmap.Size);
     }
 
-    [Theory]
+    [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsDrawingSupported))]
```
Comment (adamsitnik) [inline src/libraries/System.Resources.Extensions/tests/BinaryFormatTests/Common/SystemDrawingTests.cs:24]:
```suggestion
    [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsDrawingSupported), nameof(PlatformDetection.SupportsComInterop))]
```

**Comment by adamsitnik** [comment]
@akoeplinger thank you!!

**Comment by adamsitnik** [comment]
There is another build failure that I am not sure how to solve:

[log](https://[helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-103471-merge-19349a9639824d12b4/System.Resources.Extensions.Compat.Tests/1/console.26cf42af.log?helixlogtype=result](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-103471-merge-19349a9639824d12b4/System.Resources.Extensions.Compat.Tests/1/console.26cf42af.log?helixlogtype=result))

```log
  Xcode: /tmp/helix/working/96D20869/w/ADE9093D/e/apple_build/AppBundle/System.Resources.Extensions.Tests/System.Resources.Extensions.Tests.xcodeproj
  App: /tmp/helix/working/96D20869/w/ADE9093D/e/apple_build/AppBundle/System.Resources.Extensions.Tests/Release-appletvos/System.Resources.Extensions.Tests.app
  App moved to: /private/tmp/helix/working/96D20869/w/ADE9093D/e/System.Resources.Extensions.Tests.app/
/private/tmp/helix/working/96D20869/p/dotnet-cli/sdk/9.0.100-preview.5.24307.3/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/private/tmp/helix/working/96D20869/w/ADE9093D/e/publish/ProxyProjectForAOTOnHelix.proj]
/private/tmp/helix/working/96D20869/p/dotnet-cli/sdk/9.0.100-preview.5.24307.3/Microsoft.Common.CurrentVersion.targets(888,5): warning MSB3539: The value of the property "BaseIntermediateOutputPath" was modified after it was used by MSBuild which can lead to unexpected build results. Tools such as NuGet will write outputs to the path specified by the "MSBuildProjectExtensionsPath" instead. To set this property, you must do so before Microsoft.Common.props is imported, for example by using Directory.Build.props.  For more information, please visit https://go.microsoft.com/fwlink/?linkid=869650 [/private/tmp/helix/working/96D20869/w/ADE9093D/e/publish/ProxyProjectForAOTOnHelix.proj]
Signing bundle /tmp/helix/working/96D20869/w/ADE9093D/e/System.Resources.Extensions.Compat.Tests.app
No embedded provisioning profile found at /tmp/helix/working/96D20869/w/ADE9093D/e/System.Resources.Extensions.Compat.Tests.app/embedded.mobileprovision! Failed to sign the app!
```

Since the goal of `System.Resources.Extensions.Compat.Tests` is to just ensure that it's possible to enable BinaryFormatter and run the tests for it, we can simply disable the whole test projects for the mobile targets

@akoeplinger But how can I do that?

**Comment by simonrozsival** [comment]
@adamsitnik you should be able to skip the whole test project by adding `<ProjectExclusions>` in [src/libraries/tests.proj](https://github.com/dotnet/runtime/blob/main/src/libraries/tests.proj)

**Review by simonrozsival — APPROVED** [review]

**Pull request merged**